### PR TITLE
Added a Nix flake containing everything needed to build, dev and run …

### PR DESCRIPTION
**Pull request by MichaelFisher1997**
This is a Nix Flake for NixOS and anyone with nix package manager installed on their system.

It allow people to enter into the flake dev shell as followed:
`nix develop` and it will install all the listed dependencies to build and run redot avoiding the pain of having to workout what dependencies you need to install and what missing.

### CODE REVIEW

**Comment by MichaelFisher1997** [comment]
might need a get job thats a container with nix that can do the following

```
nix develop
scons platform=linuxbsd
```

**Comment by MichaelFisher1997** [comment]
closed as theoparis still links to this as a reference but that one should be used going forward

**Comment by SkogiB** [comment]
We working with this one now? We are not sending PRs up to Godot by default at this time, due to a reassessment in the core team, so we can start looking at stuff like this soon.

**Comment by MichaelFisher1997** [comment]
I spoke to McDubh about adding this one in for now a few days ago so i reopened it.
A lot of people have been asking for it.
I know this one works for NixOS and Nix on other linux distros, the PR that expands on this, has support for things like Mac hardware and Arm so that can be added in a little later down the road.
(I do actually have a couple people With Macs that said they will help and test it)

But yeah theirs no rush on it

**Comment by phanirithvij** [comment]
just a suggestion, you can rebase instead of adding all these merge commits.

**Comment by Arctis-Fireblight** [comment]
Hi @MichaelFisher1997 ,
I am a relatively new maintainer here, and am going back through all of the old PRs.
I see there are a few PRs here related to Nix packaging. While I am somewhat familiar with what Nix is, this is not my area of expertise.
Would you mind explaining your intended use case with this to better help myself and the rest of the team better understand what the benefits of providing upstreamed package definitions?
We are looking to make a final decision regarding this trio of PRs shortly. ( #954 , #247 , #154 )

**Comment by MichaelFisher1997** [comment]
The flake I opened last year isn’t a full dev setup yet. Right now you can nix develop and run scons manually, it pulls in all the dependencies so the build is reproducible and avoids the usual dependency issues. It still needs updating so nix build .#build (or nix run .#build) will do the full build in one step. At the time quite a few people seemed interested, but I’ve not heard much since. I’ve now got a new PC that can actually handle compiling it properly, so I can revisit this and finish it off if there’s still interest.

**Comment by Arctis-Fireblight** [comment]
Hey @MichaelFisher1997,
I'm going over the last of the open PRs.
Can you tell me at a high level the differences between this PR vs #954 ?

I am still a little on the fence about incorporating Nix packaging into the main repo, but we probably need to come to a final decision one way or another here soon, along with which of the two PRs will get merged if either of them.

**Comment by Arctis-Fireblight** [comment]
@coderabbitai review

**Comment by MichaelFisher1997** [comment]
Hey @Arctis-Fireblight,

Great question! Let me clarify both points:

## What Nix is for here

Nix (the package manager, not just NixOS the operating system) provides **reproducible, declarative dependency management and builds**. The key benefits for Redot are:

1. **Dependency Replication**: Nix guarantees that if you have the same flake.nix inputs, you get exactly the same build environment—same library versions, same toolchain versions. This eliminates "works on my machine" issues.

2. **Zero Manual Dependency Setup**: Users can run `nix develop` and Nix will automatically download and configure all build tools (gcc, python3, pkg-config, mono, dotnet, etc.) and runtime libraries (vulkan, X11, wayland, pulseaudio, etc.) in an isolated environment. No more hunting for missing packages or dealing with version conflicts.

3. **Reproducible Builds**: With Nix, the build output is deterministic. Given the same inputs, you get bit-for-bit identical binaries—critical for security and trust.

4. **Cross-Distro Support**: Nix works on any Linux distro (Arch, Ubuntu, Fedora, NixOS, etc.) and even macOS. The same flake definition provides a consistent development experience everywhere.

**Important note**: Having a Nix dev shell doesn't mean everyone is forced to use it. It's an optional tool—contributors can still use their system package managers or manual installs if they prefer. However, using Nix is good practice because it ensures a consistent environment that matches what others (and CI) use.

## Differences between #154 and #954, and why I recommend #154

| Aspect | PR #154 (Simple) | PR #954 (Comprehensive) |
|--------|------------------|-------------------------|
| **Scope** | Dev shell only (dependencies for building) | Dev shell + full packaging for distribution |
| **Purpose** | Help developers get the right dependencies quickly | Build and distribute Redot editor + export templates via Nix |
| **What it provides** | All deps needed to run `scons platform=linuxbsd` manually | One-command builds: `nix build .#redot`, export templates, CI-ready derivations |
| **Complexity** | Simple (~71 lines in flake.nix) | Complex (package.nix, export-template-package.nix, overlays) |
| **Maintenance burden** | Minimal—static list of deps, rarely needs updates | Higher—must track upstream Nix packaging standards, lockfiles, cross-compilation issues |

**My recommendation: Merge #154, move #954 to a separate repo**

I believe the Redot engine repository should focus on **source code and tooling directly needed to build it**. PR #154 fits this scope perfectly—it's just a dev shell that gives contributors a reproducible environment with all the dependencies they need to build Redot.

PR #954, on the other hand, is about **packaging and distribution**—that's a different concern. It adds 356 lines of packaging definitions, overlays, export template derivations, etc. This belongs in a dedicated Nix packaging repository that can be maintained independently by the Nix community, similar to how the official Nixpkgs repository handles thousands of packages.

**Benefits of this separation:**
- **Smaller surface area** for maintainers to review in this repo
- **Clearer responsibilities**: main repo = build/dev tools; separate repo = distribution/packaging
- **Lower maintenance burden** on the Redot team for Nix-specific packaging issues
- PR #154 is intentionally simple—just a list of dependencies that mirror what the SConstruct needs
- **Optional tooling**: No one is forced to use Nix, but those who want an easy, reproducible setup can

The Nix community already maintains packaging for many projects in external repos. @Naxdy could easily fork PR #954 into a standalone flake that references this repo. Users who want the full Nix packaging experience can use that, while anyone wanting to contribute to Redot just needs `nix develop` from #154.

**Bottom line**: Let's merge #154 to give contributors an easy, reproducible way to get dependencies. The comprehensive packaging work belongs elsewhere.

does this make sense

**Comment by MichaelFisher1997** [comment]
I've updated the branch with some improvements:

1. **Fixed nixpkgs URL** - Changed from FlakeHub to canonical GitHub URL (`github:NixOS/nixpkgs/nixpkgs-unstable`) as suggested by CodeRabbit
2. **Removed duplicate dependencies** - Cleaned up the deps list (mono, wayland-scanner, speechd, dotnet-sdk_8 were listed twice)
3. **Added one-command build/run** - You can now simply run `nix run .` to build and start Redot (or use `nix develop` for manual control)
4. **Updated README** - Added comprehensive Nix build instructions as the recommended method

**Confirmed working**: I've tested this on NixOS with `nix run .` - it builds and runs Redot successfully with all dependencies provided by Nix.

The branch is up to date with master and ready to merge when you're ready to make a decision on the Nix-related PRs.

Thanks!

**Comment by Arctis-Fireblight** [comment]
Thanks @MichaelFisher1997 !
That makes perfect sense, and honestly, better aligns with project goals.
If this is ready to go and passes the checks, I feel comfortable merging this, but it needs to go into Dev and not master.

Do you mind rebasing this branch for me real quick?

**Comment by MichaelFisher1997** [comment]
@Arctis-Fireblight Done! Rebased onto `dev` - clean linear history with 8 commits, no more merge commits. Ready for review/merge when CI passes.

**Comment by MichaelFisher1997** [comment]
@coderabbitai Thanks for the feedback! I've addressed both issues in commit 312cbc0656:

**1. Hard-coded architecture fix:**
- Added dynamic architecture detection in `flake.nix` using `pkgs.stdenv.hostPlatform.isAarch64`
- The binary name now correctly resolves to `arm64` on ARM64 systems (Apple Silicon, ARM Linux) and `x86_64` on x86_64 systems

**2. Platform target mismatch fix:**
- Updated README manual instructions to show both platform options (`linuxbsd` for Linux, `macos` for macOS)
- Added a note explaining that `nix run .` automatically detects platform and architecture

The flake now correctly supports all four system types:
- `x86_64-linux` → `redot.linuxbsd.editor.x86_64`
- `aarch64-linux` → `redot.linuxbsd.editor.arm64`
- `x86_64-darwin` → `redot.macos.editor.x86_64`
- `aarch64-darwin` → `redot.macos.editor.arm64`
